src/node_http2.cc

```diff
@@ -199,27 +210,166 @@ Http2Priority::Http2Priority(Environment* env,
+      nva[0].name = nva[0].value = &zero;
+      nva[0].namelen = nva[0].valuelen = 1;
+      count_ = 1;
+      return;
+    }
+
+    nva[n].flags = NGHTTP2_NV_FLAG_NONE;
+    nva[n].name = reinterpret_cast<uint8_t*>(p);
+    nva[n].namelen = strlen(p);
+    p += nva[n].namelen + 1;
+    nva[n].value = reinterpret_cast<uint8_t*>(p);
+    nva[n].valuelen = strlen(p);
+    p += nva[n].valuelen + 1;
+  }
+}
+
+
+Http2Session::Callbacks::Callbacks(bool kHasGetPaddingCallback) {
+  CHECK_EQ(nghttp2_session_callbacks_new(&callbacks), 0);
+  nghttp2_session_callbacks_set_on_begin_headers_callback(
+    callbacks, OnBeginHeadersCallback);
+  nghttp2_session_callbacks_set_on_header_callback2(
+    callbacks, OnHeaderCallback);
+  nghttp2_session_callbacks_set_on_frame_recv_callback(
+    callbacks, OnFrameReceive);
+  nghttp2_session_callbacks_set_on_stream_close_callback(
+    callbacks, OnStreamClose);
+  nghttp2_session_callbacks_set_on_data_chunk_recv_callback(
+    callbacks, OnDataChunkReceived);
+  nghttp2_session_callbacks_set_on_frame_not_send_callback(
+    callbacks, OnFrameNotSent);
+  nghttp2_session_callbacks_set_on_invalid_header_callback2(
+    callbacks, OnInvalidHeader);
+  nghttp2_session_callbacks_set_error_callback(
+    callbacks, OnNghttpError);
+
+  if (kHasGetPaddingCallback) {
+    nghttp2_session_callbacks_set_select_padding_callback(
+      callbacks, OnSelectPadding);
+  }
+}
+
+
+Http2Session::Callbacks::~Callbacks() {
+  nghttp2_session_callbacks_del(callbacks);
+}
+
+
 Http2Session::Http2Session(Environment* env,
                            Local<Object> wrap,
                            nghttp2_session_type type)
     : AsyncWrap(env, wrap, AsyncWrap::PROVIDER_HTTP2SESSION),
-      StreamBase(env) {
+      session_type_(type) {
   MakeWeak<Http2Session>(this);
 
   Http2Options opts(env);
 
+  int32_t maxHeaderPairs = opts.GetMaxHeaderPairs();
+  max_header_pairs_ =
+      type == NGHTTP2_SESSION_SERVER
+          ? std::max(maxHeaderPairs, 4)     // minimum # of request headers
+          : std::max(maxHeaderPairs, 1);    // minumum # of response headers
+
+  max_outstanding_pings_ = opts.GetMaxOutstandingPings();
+
   padding_strategy_ = opts.GetPaddingStrategy();
 
-  int32_t maxHeaderPairs = opts.GetMaxHeaderPairs();
-  maxHeaderPairs = type == NGHTTP2_SESSION_SERVER ?
-      std::max(maxHeaderPairs, 4) : std::max(maxHeaderPairs, 1);
-  Init(type, *opts, nullptr, maxHeaderPairs);
+  bool hasGetPaddingCallback =
+      padding_strategy_ == PADDING_STRATEGY_MAX ||
+      padding_strategy_ == PADDING_STRATEGY_CALLBACK;
+
+  nghttp2_session_callbacks* callbacks
+      = callback_struct_saved[hasGetPaddingCallback ? 1 : 0].callbacks;
+
+  typedef int (*init_fn)(nghttp2_session** session,
+                         const nghttp2_session_callbacks* callbacks,
+                         void* user_data,
+                         const nghttp2_option* options);
+  init_fn fn = type == NGHTTP2_SESSION_SERVER ?
```

Comment:
mhhhhh… `auto`? :smile:

Reply:
yeah yeah... ;-) I haven't quite managed to get used to using `auto` too much. I'm working on it, I swear ;-)